File: binary_hash.py
__author__ = 'feng'

import string
import random
import time
import numpy as np
import matplotlib.pyplot as plt


def binary_search(arr, key):
    lo, hi = 0, (len(arr) - 1)
    while lo <= hi:
        mid = (lo + hi) >> 1
        k = arr[mid]
        if k == key:
            return mid
        elif k < key:
            lo = mid + 1 # search upper subarray
        else:
            hi = mid - 1 # search lower subarray
            # (-(insertion point) - 1)
    return -(lo + 1)


def random_string(size, chars=string.digits + string.ascii_letters):
    return ''.join(random.choice(chars) for x in range(size))


COUNT = 1000 * 200
KEY_COUNTS = list(range(2000, COUNT + 1, COUNT / 10))
LOOKUP = 1500


def compare_dict_bs(data, name, array_type=None):
    all_keys = data.keys()
    items = data.items()

    rets = []
    for count in KEY_COUNTS:
        map = dict(items[:count])
        keys_arr = sorted(map.keys())

        lookup_keys = [random.choice(keys_arr) for i in range(LOOKUP / 2)] + \
                      [random.choice(all_keys) for i in range(LOOKUP / 2)]

        if array_type:
            import array

            keys_arr = array.array(array_type, keys_arr)

        map_counter = 0
        start = time.time()
        for key in lookup_keys:
            if map.get(key, None):
                map_counter += 1
        map_time = time.time() - start

        bs_counter = 0
        start = time.time()
        for key in lookup_keys:
            if binary_search(keys_arr, key) >= 0:
                bs_counter += 1

        bs_time = time.time() - start

        if bs_counter != map_counter:
            raise Exception("hash map and binary search's result conflict")

        print name, count, map_time, bs_time
        rets.append((count, map_time * 1000, bs_time * 1000))
    return rets


def plot_result(numbers, name, output):
    ind = np.arange(len(numbers))
    width = 0.35

    fig = plt.figure(figsize=(7, 5))
    ax = fig.add_subplot(111)

    rects1 = ax.bar(ind, [mt for c, mt, bt in numbers], width, color='r')
    binary_times = [bt for c, mt, bt in numbers]
    rects2 = ax.bar(ind + width, binary_times, width, color='y')

    ax.legend((rects1[0], rects2[0]), ('dict', "binary-search"))

    ax.set_ybound(upper=max(binary_times) * 1.3)
    ax.set_ylabel('time (ms)')
    ax.set_xlabel('items count')
    ax.set_title('%s: dict vs binary-search, %s lookups' % (name, LOOKUP))
    ax.set_xticks(ind + width)
    ax.set_xticklabels(["%dk" % (c / 1000) for c, mt, bt in numbers])

    if output:
        plt.savefig(output)
    else:
        plt.show()


def main():
    key_str = {}
    for i in range(COUNT):
        key_str[random_string(random.randint(2, 15))] = True

    results = compare_dict_bs(key_str, "string")
    plot_result(results, 'string', "/tmp/string.png")

    key_int = {}
    for i in range(COUNT):
        key_int[random.randint(0, 1 << 32)] = True
    results = compare_dict_bs(key_int, "int(array)", array_type='I')
    plot_result(results, 'int(array)', "/tmp/int_array.png")

    results = compare_dict_bs(key_int, "int")
    plot_result(results, 'int', "/tmp/int.png")

if __name__ == "__main__":
    main()

File: binary_hash.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
	"time"
)

func binary_search0(arr []int, key int) int {
	return binary_search_(arr, 0, len(arr), key)
}

func binary_search_(arr []int, begin, end, key int) int {
	if begin > end {
		return -1
	}
	mid := (begin + end) / 2
	if arr[mid] == key {
		return mid
	} else if arr[mid] > key {
		return binary_search_(arr, begin, mid-1, key)
	} else {
		return binary_search_(arr, mid+1, end, key)
	}
}

func binary_search(arr []int, key int) int {
	lo, hi := 0, len(arr)-1
	for lo <= hi {
		mid := (lo + hi) / 2
		if arr[mid] == key {
			return mid
		} else if arr[mid] < key { // search upper subarray
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}
	return -(lo + 1)
}

func binary_search2(arr []int, key int) int {
	lo, hi := 0, len(arr)-1
	for lo < hi {
		mid := (lo + hi) / 2
		if arr[mid] < key {
			lo = mid + 1
		} else {
			hi = mid
		}
	}

	if lo == hi && arr[lo] == key {
		return lo
	} else {
		return -(lo + 1)
	}
}

type Counter struct {
	items           int
	bs_time, m_time int64
}

const (
	COUNT  = 1000 * 1000
	LOOKUP = 1500
)

func compare_dict_bs(data []int) (counters []*Counter) {
	fmt.Printf("itemcount map 2-branch 3-branch recusive library\n")
	// counters := make([]*Counter, 10)
	for count := 2000; count < len(data); count = count + len(data)/10 {
		m := make(map[int]bool, COUNT)
		arr := make([]int, count)
		copy(arr, data)

		lookup_keys := make([]int, LOOKUP)
		copy(lookup_keys, arr[:(LOOKUP/2)])
		copy(lookup_keys[LOOKUP/2:], data[:])
		for i := LOOKUP / 2; i < LOOKUP; i++ {
			lookup_keys[i] = data[rand.Int()%len(data)]
		}

		sort.Ints(arr)
		for _, i := range arr {
			m[i] = true
		}

		start := time.Now()
		bs_find := 0
		for _, key := range lookup_keys {
			if i := sort.SearchInts(arr, key); i < len(arr) && arr[i] == key {
				// if binary_search(arr, key) >= 0 {
				// fmt.Println(binary_search(arr, key))
				bs_find += 1
			}
		}
		bs_lib := time.Since(start)

		// fmt.Println("lib", bs_lib, bs_find)

		start = time.Now()
		bs2_find := 0
		for _, key := range lookup_keys {
			if binary_search(arr, key) >= 0 {
				// fmt.Println(binary_search(arr, key))
				bs2_find += 1
			}
		}
		bs2_time := time.Since(start)
		// fmt.Println("hand(3)", bs_time, bs2_find)

		start = time.Now()
		bs3_find := 0
		for _, key := range lookup_keys {
			if binary_search2(arr, key) >= 0 {
				// fmt.Println(binary_search(arr, key))
				bs3_find += 1
			}
		}
		bs3_time := time.Since(start)
		// fmt.Println("hand(2)", bs2_time, bs3_find)

		start = time.Now()
		bs4_find := 0
		for _, key := range lookup_keys {
			if binary_search0(arr, key) >= 0 {
				// fmt.Println(binary_search(arr, key))
				bs4_find += 1
			}
		}
		bs4_time := time.Since(start)
		// fmt.Println("hand(recusive)", bs4_time, bs4_find)

		// fmt.Println("binary search", count, bs_find, bs_time, bs_time.Nanoseconds())

		start = time.Now()
		m_find := 0
		for _, key := range lookup_keys {
			if e, _ := m[key]; e {
				m_find += 1
			}
		}
		m_time := time.Since(start)
		// fmt.Println("map", m_time, m_find)
		// t := &Counter{items: count, bs_time: bs_time.Nanoseconds(),
		// 	m_time: m_time.Nanoseconds()}

		fmt.Println(count, int64(m_time), int64(bs3_time), int64(bs2_time),
			int64(bs4_time), int64(bs_lib))
		// fmt.Printf("%+v\n", t)
		// counters = append(counters, t)
	}
	return
}

func main() {
	arr := make([]int, COUNT)

	for i := 0; i < len(arr); i++ {
		t := rand.Int() % 0xffffffff
		arr[i] = t
	}

	var a []interface{} = make([]interface{}, 1)

	if _, ok := a[0].(int); ok {
		fmt.Println("OK")
	}

	// switch()

	// fmt.Printf("%T, %T\n", a, a.([]int))

	// arr is a slice, is copied, slice has a pointer to array, is not copied
	compare_dict_bs(arr)
}
